src/cluster_legacy.c

```diff
@@ -3312,10 +3314,16 @@ int clusterProcessPacket(clusterLink *link) {
                     if (sender->replicaof) clusterNodeRemoveReplica(sender->replicaof, sender);
                     serverLog(LL_NOTICE, "Node %.40s (%s) is now a replica of node %.40s (%s) in shard %.40s",
                               sender->name, sender->human_nodename, sender_claimed_primary->name,
-                              sender_claimed_primary->human_nodename, sender->shard_id);
+                              sender_claimed_primary->human_nodename, sender_claimed_primary->shard_id);
                     clusterNodeAddReplica(sender_claimed_primary, sender);
                     sender->replicaof = sender_claimed_primary;
```

Comment:
This is the only place where `->replicaof` state can be updated on this path so the most natural fix would be to pull the chain replication reduction logic into this code block and before the `updateShardId` call. We don't need to introduce yet another very purpose-built state this way.

Reply:
ok, make sense to me, i will try to think of a test case that can cover it, by any chance you have the hint on the test case?

Reply:
>by any chance you have the hint on the test case?

I think we could get a consistent repro if we disable `cluster-allow-replica-migration` on `replica 1`. Something like below:

* two shards, `shard 1` with `primary 1` and `replica 2`, and `shard 2` with `primary 2`
* `shard 1` has one slot, slot 0, while `shard 2` has the remaining slots
* `cluster-allow-replica-migration` is enabled on `primary 1` 
* `cluster-allow-replica-migration` is disabled on `replica 1`

Reply:
ok, i get it, i can take a try, just kind of a reverse version of #928